src/rust/catloop/socket.rs

```diff
@@ -283,58 +289,91 @@ impl Socket {
     }
 
     /// Asynchronously closes this socket by allocating a coroutine.
-    pub fn async_close<F>(&mut self, coroutine: F, yielder: Yielder) -> Result<(QToken, bool), Fail>
+    pub fn async_close<F>(&mut self, coroutine: F, yielder: Yielder) -> Result<TaskHandle, Fail>
     where
         F: FnOnce(Yielder) -> Result<TaskHandle, Fail>,
     {
         self.state.prepare(SocketOp::Close)?;
-        if let Some(qd) = self.catmem_qd {
-            match self.catmem.borrow().async_close(qd) {
-                Ok(qt) => {
-                    self.state.commit();
-                    Ok((qt, false))
+        Ok(self.do_generic_sync_control_path_call(coroutine, yielder)?)
+    }
+
+    /// Closes `socket`.
+    pub async fn do_close(socket: Rc<RefCell<Self>>, yielder: Yielder) -> Result<(QDesc, OperationResult), Fail> {
+        // Should we assert that we're stil in the close state?
+        let qtable: Rc<RefCell<IoQueueTable<CatmemQueue>>> = socket.borrow().catmem.borrow().get_qtable().clone();
+        if let Some(qd) = socket.borrow().catmem_qd {
+            let queue: CatmemQueue = socket.borrow().catmem.borrow().get_queue(qd)?;
+            socket.borrow_mut().state.prepare(SocketOp::Closed)?;
+
+            match CatmemLibOS::close_coroutine(qtable, queue, qd, yielder).await {
+                (qd, OperationResult::Close) => {
+                    socket.borrow_mut().state.commit();
+                    Ok((qd, OperationResult::Close))
                 },
-                Err(e) => {
-                    self.state.rollback();
-                    Err(e)
+                (qd, OperationResult::Failed(e)) => {
+                    // Where to revert to?
+                    socket.borrow_mut().state.rollback();
+                    Ok((qd, OperationResult::Failed(e)))
                 },
+                _ => panic!("Should not return anything other than close or fail"),
             }
         } else {
-            Ok((
-                self.do_generic_sync_control_path_call(coroutine, yielder)?
-                    .get_task_id()
-                    .into(),
-                true,
-            ))
+            // We know that the queue descriptor will be replaced so it doesn't matter what we put here.
+            Ok((QDesc::from(0), OperationResult::Close))
```

Comment:
Let's create a constant for this and change it to the maximum value then :) It is better to read in the logs that way.